RFC Draft.md

```diff
@@ -382,9 +382,249 @@ const uploadedObjectReference = await uploadTask.result;
+};
+
+const result = await API.query<MyQueryType>('query getTodo...', {
+  variables: { ... }
+});
+
+console.log(
+  `Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`
+);
+
+type MyMutationType = {
+  variables: {
+    id: number;
+    name: string;
+    description: string;
+  };
+  result: {
+    id: number;
+    name: string;
+    description: string;
+  };
+};
+
+const result = await API.mutate<MyMutationType>('mutation createTodo....', {
+  variables: {
+    id: 123,
+    name: 'My Todo',
+    description: 'This is a todo',
+  },
+});
+
+console.log(
+  `Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`
+);
+
+API.subscribe<MyQueryType>('subscription OnCreateTodo...').on({
+  next: (result) =>
+    console.log(
+      `Todo info: ${result.data?.id}: ${result.data?.name} (${result.data?.description})`
+    ),
+});
+```
+
+## Less verbose type definitions for generated queries, mutations, and subscriptions
+In v6, we want to reduce the verbosity of the typings for the code-generated queries, mutations, and subscriptions by inferring their types from the generated code.
+
+**Current Usage (v5)**
+```ts
+import { API, graphqlOperation } from 'aws-amplify';
+import { GraphQLQuery, GraphQLSubscription } from '@aws-amplify/api';
+import { createTodo } from './graphql/mutations';
+import { onCreateTodo } from './graphql/subscriptions';
+import {
+  CreateTodoInput,
+  CreateTodoMutation,
+  OnCreateTodoSubscription,
+} from './API';
+// so many imports from disparate modules 
+
+function createMutation() {
+  const createInput: CreateTodoInput = {
+    name: 'Improve API TS support ',
+  };
+
+  // Verbose explicit type definition when the information could be available in `createTodo`'s type
+  const res = await API.graphql<GraphQLQuery<CreateTodoMutation>>(
+    graphqlOperation(createTodo, {
+      input: createInput,
+    })
+  );
+
+  // the returned data is nested 2 levels deep and could be upleveled when the GraphQL document
+  // only includes one query or mutation
+  const newTodo = res.data?.createTodo;
+
+  console.log('new todo created', newTodo);
+}
+
+function subscribeToCreate() {
+  const sub = API.graphql<GraphQLSubscription<OnCreateTodoSubscription>>(
+    graphqlOperation(onCreateTodo)
+  ).subscribe({
+    next: (message) => {
+      // once again, we could "sift up" the return value instead of providing it in two levels of depth
+      const newTodo = message.value?.data?.onCreateTodo;
+      
+      console.log('new todo created', newTodo);
+    }
+  });
+}
+```
+
+**Proposed Usage (v6)**
+```ts
+import { API, graphqlOperation } from 'aws-amplify';
+import { GraphQLQuery, GraphQLSubscription } from '@aws-amplify/api';
+import { createTodo } from './graphql/mutations';
+import { onCreateTodo } from './graphql/subscriptions';
+import {
+  CreateTodoInput,
+  CreateTodoMutation,
+  OnCreateTodoSubscription,
+} from './API';
+// so many imports from disparate modules 
```

Comment:
I think we might want to leave the imports out of the sample to direct focus onto the changes. Regardless, for this example there a few unused imports we can clean up I think.

Reply:
agree removed.